src/ngraph/node_output.hpp

```diff
@@ -0,0 +1,91 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include "ngraph/descriptor/input.hpp"
+#include "ngraph/node.hpp"
+
+namespace ngraph
+{
+    /// \brief A handle for one of a node's outputs.
+    class NodeOutput
+    {
+    public:
+        /// \brief Constructs a NodeOutput.
+        /// \param node Pointer to the node for the output handle.
```

Comment:
shared_ptr<Node>, not pointer. Ugh, shared_ptr. So the output it the tie that binds. That which holds the graph together. The asymmetry between the NodeInput constructor signatures and the output of get_node on the two classes is ugly. Symmetry is better. Any way to achieve that? 
What horrible things would happen if instead of spewing shared_ptr everywhere we used raw pointer and did things the old-fashioned way, with new and delete?

Reply:
> shared_ptr, not pointer.

True, will change :)

In general I don't much like our use of `shared_ptr` either. I have some ideas on how to remedy that, but I think it can be addressed more or less independently of this work. One thing we _could_ do, I suppose, is store a raw pointer internally and leverage `shared_from_this`. That feels a bit like trading a venial sin for a mortal one, but then again it'd get us the symmetry we want and force users of `NodeOutput` to at least choose explicitly which flavor of pointer they want.